src/shims/fs.rs

```diff
@@ -23,6 +23,17 @@ pub struct FileHandler {
     low: i32,
 }
 
+impl FileHandler {
+    fn next_fd(&self) -> i32 {
+        self.low + 1
+    }
+
+    fn register_fd(&mut self, fd: i32, handle: FileHandle) {
+        self.low = fd;
+        self.handles.insert(fd, handle).unwrap_none();
+    }
```

Comment:
Why this complicated and dangerous of an API? Now clients can mis-use it accidentally to mess up the FD bookeeping.

I was thinking of something like:
```
fn fresh_fd(&mut self) -> i32 {
  let fd = self.next_fd;
  self.next_fd = self.next_fd.checked_add(1).unwrap();
  fd
}
```
(Not sure what "low" as a name is supposed to refer to.)

Reply:
If you'd like, having a `register_fresh_fd` that *just* takes a handle and does both calling `fresh_fd` and insert it into the HashMap, that could make sense as well.

Reply:
On most systems a new fd is **guaranteed** to be the lowest free number possible. Some programs use this instead of `dup2` to open a file with a known fd, like 0, 1 or 2.

Reply:
I don't think the current implementation we have here guarantees that.

Reply:
This API was driven by having to support the third argument to fcntl with F_DUPFD, which can change what fd gets selected. I could try redesigning this to require just one method call, maybe using a closure, or maybe separate methods for the normal case vs. fcntl.

Reply:
> On many systems a new fd is **guaranteed** to be the lowest free number possible. Some programs use this instead of `dup2` to open a file with a known fd, like 0, 1 or 2.

I think that's what `dup` does:
```
The dup() system call creates a copy of the file descriptor oldfd,
using the lowest-numbered unused file descriptor for the new
descriptor.
```
`fcntl` only uses the smallest available larger than the current one:
```
Duplicate the file descriptor fd using the lowest-numbered
available file descriptor greater than or equal to arg. 
```
Do we know of (someone)[https://xkcd.com/1172/] actually using the fact that is the "smallest" fd? In that case I think the best would be to remove the counter field from `FileHandler` and just start looking every possible fd until one is available.

Another option would be to have a `Vec` with the available file descriptors. So when we close a file, we add it to this vector, and when we open a file, we check if the next is free and add it to the vector too.

Reply:
> This API was driven by having to support the third argument to fcntl with F_DUPFD, which can change what fd gets selected

I didn't realize this was the case, makes sense.

The API seems dangerous though as it now fails to ensure that `next_fd` is actually a fresh FD.